internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -0,0 +1,585 @@
+/*
+Copyright 2022 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+this file except in compliance with the License. You may obtain a copy of the
+License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed
+under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+package composite
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/structpb"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/composed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/composite"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/fn/proto/v1beta1"
+)
+
+// Error strings.
+const (
+	errFetchXRConnectionDetails = "cannot fetch composite resource connection details"
+	errGetExistingCDs           = "cannot get existing composed resources"
+	errBuildObserved            = "cannot build observed state for RunFunctionRequest"
+	errBuildDesired             = "cannot build desired state for RunFunctionRequest"
+	errGarbageCollectCDs        = "cannot garbage collect composed resources that are no longer desired"
+	errApplyXR                  = "cannot apply composite resource"
+	errAnonymousCD              = "encountered composed resource without required \"" + AnnotationKeyCompositionResourceName + "\" annotation"
+	errUnmarshalDesiredXR       = "cannot unmarshal desired composite resource from RunFunctionResponse"
+	errXRAsStruct               = "cannot encode composite resource to protocol buffer Struct well-known type"
+
+	errFmtResourceName               = "composed resource %q"
```

Comment:
Thanks! I've noticed a super weird thing with my editor/linter in that _sometimes_ it will notice these unused constants but other times it won't. I swear it sometimes notices only when I start messing with the `const` block.